src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java

```diff
@@ -21,85 +22,93 @@
-        } else {
+        if (workspace != null) {
             LOGGER.fine("Using existing workspace: " + workspace.getName());
-            WorkspaceManager.cleanWorkspace(tool, workspace.getPath());
+            WorkspaceManager.cleanWorkspace(tool, workspace.getPath(), cleanup);
+        } else {
+            String workspaceName = WorkspaceManager.generateUniqueWorkspaceName();
+            LOGGER.fine("Creating new workspace: " + workspaceName);
+            if (workspacePath.exists()) {
+                workspacePath.deleteContents();
+            }
+            workspace = WorkspaceManager.createWorkspace(tool, workspacePath, workspaceName, selector);
         }
 
-        WorkspaceManager.setWorkspaceSelector(tool, workspacePath, selector);
+        LOGGER.fine("Changing workspace selector to '" + StringUtil.singleLine(selector) + "'");
+        WorkspaceManager.setSelector(tool, workspacePath, selector);
+        // Setting workspace selector triggers workspace update
+        // WorkspaceManager.updateWorkspace(tool, workspace.getPath());
 
         return workspace;
     }
 
-    private static void cleanOldWorkspacesIfNeeded(
-            PlasticTool tool,
-            FilePath workspacePath,
-            boolean shouldUseUpdate,
-            List<Workspace> workspaces) throws IOException, InterruptedException {
+    private static boolean mustUpdateSelector(PlasticTool tool, FilePath workspacePath, String selector)
+            throws IOException, InterruptedException, ParseException {
+        String actualSelector = StringUtil.removeNewLines(WorkspaceManager.getSelector(tool, workspacePath));
+        String expectedSelector = StringUtil.removeNewLines(selector);
+        return !actualSelector.equals(expectedSelector);
+    }
 
-        // handle situation where workspace exists in parent path
-        Workspace parentWorkspace = findWorkspaceByPath(workspaces, workspacePath.getParent());
-        if (parentWorkspace != null) {
-            deleteWorkspace(tool, parentWorkspace, workspaces);
-        }
+    private static void deleteOldWorkspacesIfNeeded(PlasticTool tool, FilePath workspacePath, CleanupMethod cleanup, List<Workspace> workspaces)
+            throws IOException, InterruptedException {
 
-        // handle situation where workspace exists in child path
-        List<Workspace> nestedWorkspaces = findWorkspacesInsidePath(workspaces, workspacePath);
-        for (Workspace workspace : nestedWorkspaces) {
+        // Handle situation where workspace exists in child path.
+        List<Workspace> innerWorkspaces = findWorkspacesInsidePath(workspacePath, workspaces);
+        for (Workspace workspace : innerWorkspaces) {
             deleteWorkspace(tool, workspace, workspaces);
         }
 
-        if (shouldUseUpdate) {
-            return;
+        // Handle situation where workspace exists in parent path.
+        List<Workspace> outerWorkspaces = findWorkspacesOutsidePath(workspacePath, workspaces);
```

Comment:
I think a better name for this would be `parentWorkspaces`, right?

Reply:
Parent would indicate it is the immediate directory above (1 level up). This list will however gather all workspaces which are above this one (all levels up).

Reply:
We'll accept this change. We'll think if we can come up with a better naming 😉